pkg/controller/bucketprovisioner/noobaa_bucket_provisioner.go

```diff
@@ -0,0 +1,331 @@
+	}
+	logger.Infof("Successfully created bucket %q", p.bucketName)
+	return nil
+}
+
+func (p *noobaaBucketProvisioner) createAccountForBucket() error {
+	timeSuffix := time.Now().Unix()
+	name := fmt.Sprintf("%s.account-%v@noobaa.io", p.bucketName, timeSuffix)
+	accountInfo, err := p.nbClient.CreateAccountAPI(nb.CreateAccountParams{
+		Name:              name,
+		Email:             name,
+		HasLogin:          false,
+		S3Access:          true,
+		AllowBucketCreate: false,
+		AllowedBuckets: nb.AccountAllowedBuckets{
+			FullPermission: false,
+			PermissionList: []string{p.bucketName},
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	if len(accountInfo.AccessKeys) == 0 {
+		return fmt.Errorf("Create account did not return access keys")
+	}
+
+	p.accountUserName = name
+	p.accessKey = accountInfo.AccessKeys[0].AccessKey
+	p.secretKey = accountInfo.AccessKeys[0].SecretKey
+	logger.Infof("Successfully created account %q with access to bucket %q", name, p.bucketName)
+	return nil
+}
+
+func (p *noobaaBucketProvisioner) deleteAccount() error {
+
+	logger.Infof("deleting account %q", p.accountUserName)
+	_, err := p.nbClient.DeleteAccountAPI(nb.DeleteAccountParams{Email: p.accountUserName})
+	if err != nil {
+		return fmt.Errorf("failed to delete account %q. got error: %v", p.accountUserName, err)
+	}
+	return nil
+}
+
+func (p noobaaBucketProvisioner) Provision(options *apibkt.BucketOptions) (*obAPI.ObjectBucket, error) {
+	p.bucketName = options.BucketName
+	logger.Infof("Provision: got request to provision bucket %q", p.bucketName)
+
+	err := p.initNoobaaInfo()
+	if err != nil {
+		logger.Info("GetNBClient returned error ", err)
+		return nil, err
+	}
+
+	// TODO: we need to better handle the case that a bucket was created, but Provision failed
+	// right now we will fail on create bucket when Provision is called the second time
+	err = p.createBucket()
+	if err != nil {
+		return nil, err
+	}
+
+	// create account and give permissions for bucket
+	p.createAccountForBucket()
+	if err != nil {
+		return nil, err
+	}
+
+	return p.getObjectBucket(), nil
+}
+
+func (p noobaaBucketProvisioner) Delete(ob *obAPI.ObjectBucket) error {
+	p.bucketName = ob.Spec.Endpoint.BucketName
+	p.accountUserName = ob.Spec.AdditionalState[obStateAccountNameKey]
+	logger.Infof("Delete: got request to delete bucket %q and account %q", p.bucketName, p.accountUserName)
+
+	err := p.initNoobaaInfo()
+
+	// TODO: delete the bucket and all its object
+
+	err = p.deleteAccount()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p noobaaBucketProvisioner) Grant(options *apibkt.BucketOptions) (*obAPI.ObjectBucket, error) {
+	p.bucketName = options.BucketName
+	logger.Infof("Grant: got request to grant access to bucket %q", p.bucketName)
+
+	err := p.initNoobaaInfo()
+	if err != nil {
+		logger.Info("GetNBClient returned error ", err)
+		return nil, err
+	}
+
+	// create account and give permissions for bucket
+	p.createAccountForBucket()
```

Comment:
need to check bucket exists

Reply:
right now it fails anyway. I will fix it in next PR to fail more gracefully